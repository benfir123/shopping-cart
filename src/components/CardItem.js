import { useEffect } from "react";

const CardItem = ({
  name,
  image,
  price,
  handleChange,
  handleIncrement,
  handleDecrement,
  handleAddToCart,
  quantities,
}) => {
  const handleIncrementWithParam = handleIncrement.bind(null, name);
  const handleDecrementWithParam = handleDecrement.bind(null, name);
  const handleAddToCartWithParam = handleAddToCart.bind(null, name);

  useEffect(() => {
    const incrementButton = document.querySelector(
      `[listid=${name.replace(/\s/g, "")}] > #increment`
    );
    const decrementButton = document.querySelector(
      `[listid=${name.replace(/\s/g, "")}] > #decrement`
    );
    const addToCartButton = document.querySelector(
      `[listid=${name.replace(/\s/g, "")}] > #add-to-cart`
    );

    incrementButton.addEventListener("click", handleIncrementWithParam);
    decrementButton.addEventListener("click", handleDecrementWithParam);
    addToCartButton.addEventListener("click", handleAddToCartWithParam);

    return () => {
      incrementButton.removeEventListener("click", handleIncrementWithParam);
      decrementButton.removeEventListener("click", handleDecrementWithParam);
      addToCartButton.removeEventListener("click", handleAddToCartWithParam);
    };
  });

  return (
    <div className="card-item" listid={name.replace(/\s/g, "")}>
      <img src={image} alt="" />
      <h4>{name}</h4>
      <p>{price}</p>
      <button id="decrement">-</button>
      <input
        type="number"
        name="quantity"
        id=""
        min="0"
        value={quantities.find((item) => item.name === name).quantity}
        onChange={handleChange(name)}
      />
      <button id="increment">+</button>
      <button id="add-to-cart">Add To Cart</button>
    </div>
  );
};

export default CardItem;
