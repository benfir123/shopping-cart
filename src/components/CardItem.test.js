import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CardItem from "./CardItem";

describe("Card item", () => {
  it("renders correct product title", () => {
    const { container } = render(<CardItem />);
    expect(container).toMatchSnapshot();
  });

  it("increments and decrements properly", () => {
    render(<CardItem />);
    const incrementButton = screen.getByRole("button", { name: "+" });
    const decrementButton = screen.getByRole("button", { name: "-" });

    userEvent.click(incrementButton);
    userEvent.click(decrementButton);

    expect(screen.getByRole("input").value).toMatch(0);
  });
  it("accepts manually typing in the quantity", () => {
    const onChangeMock = jest.fn();
    render(<CardItem onChange={onChangeMock} />);
    const input = screen.getByRole("input");

    userEvent.type(input, "123");

    expect(onChangeMock).toHaveBeenCalledWith(1, 1);
    expect(onChangeMock).toHaveBeenCalledWith(2, 12);
    expect(onChangeMock).toHaveBeenCalledWith(3, 123);
  });
  it("adds to cart after button click", () => {
    const onClickMock = jest.fn();
    render(<CardItem onClick={onClickMock} />);
    const button = screen.getByRole("button", { name: "Add To Cart" });

    userEvent.click(button);

    expect(onClickMock).toHaveBeenCalled();
  });
});
