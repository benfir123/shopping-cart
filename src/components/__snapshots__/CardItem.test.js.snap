// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card item renders correct product title 1`] = `
<div>
  <div
    class="card-item"
  >
    <img
      alt=""
      src=""
    />
    <h4>
      Title goes here
    </h4>
    <p>
      $299
    </p>
    <button>
      -
    </button>
    <input
      id=""
      name="quantity"
      type="number"
    />
    <button>
      +
    </button>
    <button>
      Add To Cart
    </button>
  </div>
</div>
`;
